### PR TITLE
Add imputation into estimate_missing script

### PR DESCRIPTION
**Pull request by garypolzin**
# Description
Imputation now outputs one column which is a combination of known data, extrapolated and interpolated data - the column name created contains the columns on the data used to create it.

Function called twice
- once for NR using filled posts and the rolling average model
- once for CH using ratio and the ratio rolling average model

# How to test
Data check running in branch - compare to previous outputs.
Check what happens to locations who switch between CH and NR.

[Trello ticket](https://trello.com/c/o0jObcV4/802-replace-current-extrap-interp-with-new-combined-version-in-the-pipeline)

# Developer checklist
- [X] Unit test
- [X] Linked to Trello ticket
- [X] Documentation up to date

### CODE REVIEW

File: tests/test_file_data.py [hunk 1]
```diff
@@ -4881,17 +4881,91 @@ class ModelPrimaryServiceRollingAverage:
...
+        ("1-002", date(2023, 2, 1), 1675209600, CareHome.not_care_home, None, 50.5),
+        ("1-002", date(2023, 3, 1), 1677628800, CareHome.not_care_home, None, 50.7),
+        ("1-002", date(2023, 4, 1), 1680303600, CareHome.not_care_home, None, 50.1),
+        ("1-003", date(2023, 1, 1), 1672531200, CareHome.not_care_home, None, 50.3),
+        ("1-003", date(2023, 2, 1), 1675209600, CareHome.not_care_home, 20.0, 50.5),
+        ("1-003", date(2023, 3, 1), 1677628800, CareHome.not_care_home, None, 50.7),
+        ("1-004", date(2023, 3, 1), 1677628800, CareHome.not_care_home, None, 50.7),
+    ]
+
+    split_dataset_rows = [
```
Comment (MarionJHolloway) [inline tests/test_file_data.py:4897]: I think there's more rows of data here than you need? You really just need 4, to cover all the permutations of the filter you are using.

File: utils/estimate_filled_posts/models/imputation_with_extrapolation_and_interpolation.py [hunk 2]
```diff
@@ -81,6 +98,70 @@ def create_imputation_model_name(
...
+
+def identify_locations_with_a_non_null_submission(
+    df: DataFrame, column_with_null_values: str
+) -> DataFrame:
+    """
+    Identifies locations with at least one non-null submission in the specified column.
+
+    This function adds a new column to the DataFrame indicating whether each location
+    has ever had a non-null value in the specified column. The new column will contain
+    True if there is at least one non-null value for the location, and False otherwise.
```
Comment (MarionJHolloway) [inline utils/estimate_filled_posts/models/imputation_with_extrapolation_and_interpolation.py:144]: Could you add in "at any point in time" just for clarity.

**Review by MarionJHolloway — CHANGES_REQUESTED** [review]
A few bits to check and reword, but generally looking really good, well done!

Don't forget to rename you're PR as well ;)

File: jobs/estimate_missing_ascwds_ind_cqc_filled_posts.py [hunk 3]
```diff
@@ -12,6 +12,9 @@
 from utils.estimate_filled_posts.models.primary_service_rolling_average import (
     model_primary_service_rolling_average,
 )
+from utils.estimate_filled_posts.models.imputation_with_extrapolation_and_interpolation import (
+    model_imputation_with_extrapolation_and_interpolation,
+)
 from utils.estimate_filled_posts.models.interpolation import model_interpolation
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply jobs/estimate_missing_ascwds_ind_cqc_filled_posts.py:18]: Yea they still need doing, separate PR for those :)

File: utils/column_names/ind_cqc_pipeline_columns.py [hunk 4]
```diff
@@ -117,7 +117,18 @@ class IndCqcColumns:
     first_rolling_average: str = "first_rolling_average"
     first_submission_time: str = "first_submission_time"
     gac_service_types: str = CQCLClean.gac_service_types
+    has_non_null_value: str = "has_non_null_value"
     import_month_index: str = "import_month_index"
+    imputation_filled_posts_rolling_average_model: str = (
+        "imputation_" + ascwds_filled_posts_dedup_clean + "_rolling_average_model"
+    )
+    imputation_posts_per_bed_ratio_rolling_average_model_filled_posts_per_bed_ratio: (
```
[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/column_names/ind_cqc_pipeline_columns.py:125]: @MarionJHolloway I don't think so unfortunately, unless we rename the original column names

interpolation includes the column name which contains nulls and the model it uses to trend against
the column name is `filled_posts_per_bed_ratio`
the model name is `rolling_average_model_filled_posts_per_bed_ratio`

later on we'll impute `filled_posts_per_bed_ratio` using `care_home_model` so can't drop the model name as won't be able to tell the difference

or we just name ourselves, but still, it's imputing the `filled_posts_per_bed_ratio` using the `rolling_average_model_filled_posts_per_bed_ratio` model

File: tests/test_file_data.py (hunk 1 above)
Comment (garypolzin) [reply tests/test_file_data.py:4897]: Reduced to 8 rows
1-001 = interpolation (needs 2 known points and one null in between)
1-002 = extrapolation (one known, one null either side)
1-003 = no data

Comment (garypolzin) [reply tests/test_file_data.py:4893]: oh I've looked at the wrong dataset!

Comment (MarionJHolloway) [reply tests/test_file_data.py:4893]: I meant the test data for splitting the dataset - there's no imputation in that function?

Comment (garypolzin) [reply tests/test_file_data.py:4893]: Take 2

Yea 4 rows is fine now - previously the 'identify_locations_with_a_non_null_submission' function was inside this one which complicated the tests but I moved it outside for simplification, but then didn't simplify the test data!

File: utils/column_names/ind_cqc_pipeline_columns.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MarionJHolloway) [reply utils/column_names/ind_cqc_pipeline_columns.py:125]: Well you have "imputation filled posts rolling average_model" for non res, so could it just be "imputation posts per bed ratio rolling average model" for care homes?

[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/column_names/ind_cqc_pipeline_columns.py:125]: there are two models, a filled post on called `rolling_average_model` and another ratio one called `rolling_average_model_filled_posts_per_bed_ratio`

[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/column_names/ind_cqc_pipeline_columns.py:125]: "imputation posts per bed ratio rolling average model" implies we've used `rolling_average_model` as the basis for imputation

[the remark this answers is not in the document]
Comment (MarionJHolloway) [reply utils/column_names/ind_cqc_pipeline_columns.py:125]: Could we dial it back further and just call them "imputation non res" and "imputation care homes"?

[the remark this answers is not in the document]
Comment (garypolzin) [reply utils/column_names/ind_cqc_pipeline_columns.py:125]: I don't think so, in the final estimates model we currently have `extrapolation_rolling_average_model` and `extrapolation_non_res_with_dormancy_model` which are both non res but different models

could probably get rid of the column with null names so it does what extrapolation does now. so it won't say whether it's imputing posts or ratio, it'll just say which model it used

**Review by MarionJHolloway — APPROVED** [review]
Looks good!

**Pull request merged**